sw/otbn/code-snippets/calc_montgomery_consts.s

```diff
@@ -0,0 +1,285 @@
+ .balign 4
+ .global compute_m0inv2to256
+ .type compute_m0inv2to256, @function
+
+compute_m0inv2to256:
+  /* As odd numbers are their own inverse mod 8, the input is taken as
+      start value for a Newton iteration (quadratic Hensel lifting)
+      doubling the precision in each step.
+     The code can be adapted to calculate -m^-1 mod 3*2^k by
+      increasing the precision of the last round of first loop to
+      96 bit, and of the following loops to 192 rsp. 384 bit. */
+  bn.mov          w1, w0  /* w0 is its own inverse with 3-bit precision */
+  /* double precision via Newton in each round from 2^1 to 2^6 bit */
+  loopi           5, 4
+    bn.mulqacc.wo.z w2, w1.0, w0.0, 0
+    bn.mulqacc.wo.z w2, w1.0, w2.0, 0
+    bn.add          w1, w1, w1
+    bn.sub          w1, w1, w2
+  /* double precision from 64 to 128 bit */
+  bn.mov          w2, w0
+  loopi           2, 3
+    bn.mulqacc.z    w1.0, w2.0, 0
+    bn.mulqacc      w1.0, w2.1, 64
+    bn.mulqacc.wo   w2, w1.1, w2.0, 64
+  bn.add          w1, w1, w1
+  bn.sub          w1, w1, w2
+  /* double precision from 128 to 256 bit */
+  loopi           2, 10
+    bn.mulqacc.z    w1.0, w0.0, 0
+    bn.mulqacc      w1.0, w0.1, 64
+    bn.mulqacc      w1.1, w0.0, 64
+    bn.mulqacc      w1.0, w0.2, 128
+    bn.mulqacc      w1.1, w0.1, 128
+    bn.mulqacc      w1.2, w0.0, 128
+    bn.mulqacc      w1.0, w0.3, 192
+    bn.mulqacc      w1.1, w0.2, 192
+    bn.mulqacc      w1.2, w0.1, 192
+    bn.mulqacc.wo   w0, w1.3, w0.0, 192
+  bn.add          w1, w1, w1
+  bn.sub          w0, w0, w1
+  ret
+ .size compute_m0inv2to256, .-compute_m0inv2to256
+
+/* Calculate r^2 mod m, where r = 2^{256 k} and 2^{256*k-1} < m < r.
+ *
+ *   Returns 4^{256*k} mod m given m with 2^{256*k-1} < m < 2^{256*k}
+ *
+ * This routine runs in constant time (for fixed k).
+ *
+ * @param[in]:   x10:    pointer to m, the most significant limb of m must be >=2^255
+ * @param[in]:   x11:    length k of m in 256-bit-limbs
+ * @param[in]:   x12:    pointer to output buffer for r^2 mod m, same size as m
+ *
+ * @param[out]:  x10:    0 if everything went ok, otherwise 1
+ *
+ * clobbered registers:    w0-w[3+k-1], x8-x10
+ * clobbered flag groups:  FG0, FG1
+ * called subroutines:     -
+ */
+  /* TODO: Montgomery squaring might be faster at some k, as soon r*2^k is known. */
+  /* TODO: If a fast long division is available, it might be faster to simply divide. */
+  /* TODO: Support for bitlength(m) not a multiple of 256, if necessary. */
+ .section .text.compute_rr_sample
+ .balign 4
+ .global compute_rr_sample
+ .type compute_rr_sample, @function
+
+compute_rr_sample:
+  /* Load m into registers starting at w3 */
+  /*  and write its 2-complement to output buffer */
+  li          x8, 3           /* -> w3 */
+  add         x9, x12, x0     /* -> output buffer */
+  bn.sub      w1, w1, w1      /* Clear w1 and carry FG0.C */
+  loop        x11, 4
+    bn.lid      x0, 0(x10++)    /* load next limp of m to w0 */
```

Comment:
nitpick:
```suggestion
    bn.lid      x0, 0(x10++)    /* load next limb of m to w0 */
```
It looks like this typo got repeated in a few more places as well.

Reply:
Should be all fixed now (and correcting wrong English is not nitpicking).